Flatten CloudFormation stacks for faster deploys

austinbyers opened the pull request:
## Background
Panther's long deployment time is considerably slowing down the development cycle. The CloudFormation is also hard to understand, it was never organized after we smashed together the CFN from dozens of repos to form the repo you see today.

Unfortunately, existing deployments will need to start over completely after this change, which is why I'm landing it before v1.

Closes: #456 
Closes: #197 

### Nested Stacks
It turns out one of the biggest performance bottlenecks are nested stacks, for two main reasons:

1. Nested stacks break parallelization - the entire nested stack has to finish before you can reference any of its outputs. Some resources take far longer to create (e.g. KMS keys), so there is a lot of time spent waiting for one or two resources before CFN can continue
1. Nested stacks require round trips to S3 (you have to upload, CFN has to download). The S3 artifacts are not pinned to a specific version, which means at the time of change set creation CFN can't know whether there will be changes until it starts evaluating the template.

There are other good reasons to reduce our reliance on nested stacks besides improved performance:

- The default [limit](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/cloudformation-limits.html) is 200 CFN stacks, which includes nested stacks. Panther is already at 36 stacks, 18% of a user's entire quota
- Reliability: nested stacks seem to get stuck in UPDATE_ROLLBACK_FAILED states that can be difficult to recover from. With mostly flat singleton stacks, the majority of a deployment can still succeed even if a small part fails, and the failing stack can be deleted and recreated individually without having to tear down an entire deployment.

## Design
I've almost completely removed nested stacks (with a few small exceptions noted below). Panther now has 12 top-level stacks:

1. `bootstrap` - All S3 buckets, KMS keys, Appsync Api, Cognito user pool, load balancer / networking. Basically, everything that takes a long time to create and/or generates random IDs that we need later. 
1. `bootstrap-gateway` - API gateway and the Python layer - "bootstrap" resources that just require an S3 bucket to exist first (which is why they can't be in the first stack). 

These two stacks are the only ones with outputs! In fact, I actually combine the outputs from these two stacks and pass them through the entire`deploy` command. With the bootstrap stacks out of the way (~4 minutes), nearly everything else can be deployed in parallel:

- CloudWatch alarms: to avoid per-stack resource limits, we generate nested stacks for each top-level template. We could deploy each alarm stack separately, but here having a top-level stack with several nested stacks makes sense - it's a collection of many autogenerated substacks. They used to be doubly nested (monitoring => alarms => substacks), I got rid of the first layer of nesting
- CloudWatch dashboards
- CloudWatch metric filters (in their own stack because the associated log groups must exist first)
- Appsync schema + resolvers (the API ID is in bootstrap but nothing else related to appsync)
- Cloud Security
- Core (stuff shared between cloudsec and log analysis - analysis-api, users-api, etc)
- web server: this is only the ECS task (the part that actually changes when web src is modified). All the web network and load balancer is in bootstrap
- Glue databases (same as before, just deployed directly instead of as a nested stack)
- Log Analysis
- Onboarding (same as before). These are nested stacks that reference the `aux` templates, so I just left them

![Stack Counts](https://user-images.githubusercontent.com/3608925/77135031-f0d4c980-6a26-11ea-93aa-a2b684fcab39.png)

## Changes

- On a fresh deploy, the invite for the first user has been moved to the very end (closing #197). This means the user prompt no longer blocks the deploy process - when you see the prompt, you're done!
- Reorganized the config file thematically instead of exposing stack details
- The CFN resources have not been modified (just moved) with a few exceptions:
    - Renamed the "access-logs" S3 bucket to "auditlogs" (per @rleighton's request). This bucket is now always created, but you can still route S3 access logs somewhere else per the config file.
    - I removed some parameters and outputs if they weren't needed
    - There are more generated CW alarms than before (e.g. SNS topics are now included)

## Testing

- Deployed to dev account, was able to login, update my profile, see the dashboard, resources, policies, etc. @kostaspap @rleighton can you deploy and make sure log processing / remediation still work as expected?

- Timed the deploy to `us-east-1` from commit `a748131` and this branch with a macbook on gigabit internet:

![Deploy Benchmark](https://user-images.githubusercontent.com/3608925/77135264-eb2bb380-6a27-11ea-945d-ca0eabee6933.png)

About twice as fast overall!

nhakmiller commented on tools/mage/deploy.go:
```diff
@@ -254,27 +298,156 @@ func uploadLayer(awsSession *session.Session, libs []string, bucket, key string)
+	}(finishedStacks)
+
+	// Wait for stacks to finish
+	for i := 1; i <= 8; i++ {
```
Not sure if there is a good/clean way to do this, but maybe you could have a counter that increases every time you call `go func` or something so that it's more clear where this 8 is coming from? My main concern is someone adding/removing a stack in the future and forgetting to update this number. Especially adding a stack would be problematic because the deploy would still finish (and probably be ok) in most cases, but in some rare cases might cause some weirdness.

austinbyers replied:
Sure, fair point. A `sync.WaitGroup` would have the same problem (you have to set the count), I'll see what I can do

austinbyers replied:
You can't update a counter in the goroutine (unless you protect it with a mutex lock)

So what I did instead is to just keep a running total of how many parallel stacks you started like this:

```go
var parallelStacks = 0

// Appsync
parallelStacks++
go func() { .... deploy .... }

// Glue
parallelStacks++
go func() { .... deploy ... }
```

This way, developers can add and remove stacks and the count will still be accurate as long as they follow that convention

nhakmiller replied:
Right it is a bit weird either way, maybe even just a comment saying "hey be sure to update this number if you add/remove any stacks to the process". Also they'll have. to update both the two numbers in the below for loop, which might not be immediately intuitive either. Ultimately its probably a rare case, I wouldn't spend too much time on it.

austinbyers commented on docs/gitbook/operations/runbooks.md:
```diff
@@ -96,9 +102,6 @@ This lambda implements the analysis API which is responsible for
 
- Failure Impact
- * Failure of this API Gateway will prevent calls to the `panther-analysis-api` lambda.
-
```
Yeah, it just seemed redundant to see in the description "invokes the `panther-analysis-api` lambda" and then in the failure impact "fails to invoke the `panther-analysis-api` lambda"

I can put it back if we like

nhakmiller replied:
No that's fine, I was just wondering if it was across the board or just in a few of them.

austinbyers replied:
Oh I may not have been consistent with them

austinbyers commented on deployments/panther_config.yml:
```diff
@@ -16,96 +16,84 @@
 
-  # Custom layer attached to every Python Lambda function for analysis and remediation.
+Setup:
+  # List of policy/rule sets to install when Panther is FIRST deployed.
```
True, I can change the comment (I just copied what we had before)

austinbyers commented:
I've redeployed and tested this change even more thoroughly:

- Integration tests
- Clicking on everything in the web app
- Checking infra alarms (there were a few, but all related to stuff I was doing in the app)

I'd say it's ready to merge, but I will wait until @rleighton or @kostaspap can verify the log functionality so that they don't have to deal with the merge conflicts

kostaspap commented on deployments/bootstrap2.yml:
```diff
@@ -0,0 +1,147 @@
+# Panther is a scalable, powerful, cloud-native SIEM written in Golang/React.
```
Can we change the name of it? It looks a bit like a typo to me :P 
Maybe make it more explicit and rename the two stacks from `bootstrap`, `bootstrap2` to:
- `bootstrap`, `pre-reqs`
- `bootstrap-stage-1`, `bootstrap-stage-2`
- `bootstrap-1`, `bootstrap-2`. 
- ...

austinbyers replied:
Sure, how about `bootstrap` and `bootstrap-gateway`?

austinbyers replied:
I went with `bootstrap-gateway` because the main resource is API gateway and the variable names in the code tend to be `gatewayStack`, etc

kostaspap replied:
works for me

kostaspap commented on deployments/bootstrap2.yml:
```diff
@@ -0,0 +1,147 @@
+    Type: AWS::Lambda::LayerVersion
+    Properties:
+      CompatibleRuntimes:
+        - python3.6
```
Not that it matters so much (this PR is massive already!) but I think it would be sufficient to have only `python-3.7` since this is what our lambda's use.

austinbyers replied:
I'm going to leave it because people could re-use this layer in their own Python Lambda functions if they wanted

kostaspap replied:
Sure

austinbyers commented:
I've updated the branch with master. #489 and #491  were big changes that modified lots of CloudFormation, I manually reviewed both PRs and applied all the same changes here, with a few small changes:

- Apparently we do have to always supply the certificate arn, even if the bootstrap stack already exists. Not sure if that change happened recently or if I just missed this before
- I found two bugs in the deploy onboarding:
    - I already had guardduty enabled, but when `deploy_onboard.go` tried to describe the onboarding stack, it didn't exist. The error message we were looking for was not accurate. I fixed this
    - The other error is due to #494 , I just commented out the `PutIntegration` call for log processing onboarding
- Simplify onboarding teardown (it was in a separate method that can now be removed)
- Fixed integration tests broken by #482 

I went through a full testing cycle again:

- deploy to empty account
- login, click on everything in the web app
- deploy again (no change)
- integration tests
- teardown

@kostaspap @rleighton I'm going to need help testing this branch to make sure log processing still works end-to-end. There have been lots of CFN migrations which means I may have missed something.

kostaspap commented on tools/mage/deploy_onboard.go:
```diff
@@ -54,38 +51,38 @@ func genLogProcessingLabel(awsSession *session.Session) string {
-	registerPantherAccount(awsSession, bucketOutputs, backendOutputs) // this MUST follow the CloudSec roles being deployed
-	deployRealTimeStackSet(awsSession, backendOutputs["AWSAccountId"])
+func deployOnboard(awsSession *session.Session, settings *config.PantherConfig, accountID string, bootstrapOutputs map[string]string) {
+	deployOnboardTemplate(awsSession, settings, bootstrapOutputs)
```
I made the same comment to @rleighton and he recommend it that we fix the issue in this PR: We should not deploy just one template for the logs we are onboarding, we should deploy one CFN template per source: 
- One template for compliance
- One template per log type we onboard. 

This is the standard we are following when users onboard logs through the UI, we should do it here as well. This will make possible to also off-board individual sources if someone wants to.

austinbyers replied:
That's too big a change for me to make, especially since I'm not familiar with the onboarding flow. If you want to make that change to this branch, please do

austinbyers commented on deployments/panther_config.yml:
```diff
@@ -16,96 +16,85 @@
+    - https://github.com/panther-labs/panther-analysis/releases/latest/download/panther-analysis-all.zip
+
+  # Whether or not the Panther deployment should automatically onboard itself as a data source.
+  OnboardSelf: true
```
I'm just copying the same functionality we had before, changes like that should ideally not be part of this PR, but I will do so this time given the launch today.

- I agree with having it on by default because we want a quick time to value. Users will need to onboard some kind of log data to get a feel for the product
- I can add more comments to explain what is onboarded
- I'll log a warning during the onboarding that costs may be higher
- I'll update the "cost" section of the docs

austinbyers replied:
Chatted offline, we'll also disable CloudTrail and GuardDuty by default, since they scan the entire account, not just the Panther deployment region

austinbyers commented on deployments/panther_config.yml:
```diff
@@ -16,96 +16,85 @@
-# use the BulkUpload functionality from the web app to upload new or modified rule sets.
-InitialAnalysisSets:
-  - https://github.com/panther-labs/panther-analysis/releases/latest/download/panther-analysis-all.zip
+  FargateTaskMemory: 1024
```
Sure, I can remove
